src/category_theory/zigzag_2.lean

```diff
@@ -64,9 +64,28 @@ def above {n m : Δ} (f : n ⟶ m) (j : fin m) := { i : fin n | f i ≥ j }
 
 def T : Δ ⥤ Δ :=
 { obj := λ n, (n + 1 : ℕ),
-  map := λ n m f, ⟨λ i, if h : i.val < n then (f (i.cast_lt h)).cast_succ else fin.last _, sorry⟩,
-  map_id' := sorry,
-  map_comp' := sorry, } -- see https://leanprover.zulipchat.com/#narrow/stream/113488-general/topic/.60split_ifs.60.2C.20and.20nested.20.60dite.60/near/167593063
+  map := λ n m f,
+  ⟨λ i, if h : i.val < n then (f (i.cast_lt h)).cast_succ else fin.last _,
+  λ a b h,
+  begin
+    tidy,
+    split_ifs,
+    {tidy},
+    {apply fin.le_last},
+    {rw [fin.le_iff_val_le_val] at h,
+    dsimp [(Δ)] at n, -- without this line linarith doesn't know that n : ℕ and fails
```

Comment:
Possible side project: investigate if there's a suitable place to add `whnf` in the part of `linarith` that is inspecting hypotheses, and see if that is enough to correctly identify a hypothesis like this.

Even if so, and `whnf` is reasonably fast, it's not totally clear whether it would be accepted. The usual consensus is against automation like this that is potentially slow. One compromise would be a `linarith!` modifier, that says "work harder".

Reply:
In any case, this proof is sufficiently horrible that it should be broken out into a separate lemma. Try replacing all the `tidy` with `tidy?` to see what they are doing.